### PR TITLE
Migrate PoolManager Test to foundry

### PR DESCRIPTION
**Pull request by sm-stack**
## Related Issue
This pull request migrates PoolManager tests from HardHat to Foundry, a part of https://github.com/Uniswap/v4-core/issues/315 .

## Description of changes
Added 4 test cases around `take` at PoolManager.t.sol, and added 1 test cases with `testCollectFeesForNativeTokens` at Fees.t.sol. Deleted `PoolManager.spec.ts`.

### CODE REVIEW

**Comment by sm-stack** [comment]
@ewilz Could I get a review or feedback about this?

File: test/foundry-tests/utils/MockInvalidERC20.sol [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.20;
+
+import {ERC20} from "solmate/tokens/ERC20.sol";
+
+contract MockInvalidERC20 is ERC20 {
```
Comment (zhongeric) [inline test/foundry-tests/utils/MockInvalidERC20.sol:6]: We can just use `TestInvalidERC20` which is in the top level `contracts/test` folder right?

File: test/foundry-tests/Fees.t.sol [hunk 2]
```diff
@@ -581,6 +594,47 @@ contract FeesTest is Test, Deployers, TokenFixture, GasSnapshot {
...
+        swapRouter.swap(
+            key0,
+            IPoolManager.SwapParams(false, 10000, TickMath.MAX_SQRT_RATIO - 1),
+            PoolSwapTest.TestSettings(true, true),
+            ZERO_BYTES
+        );
+
+        uint256 expectedProtocolFees = 3; // 10% of 30 is 3
+        vm.prank(address(protocolFeeController));
+        manager.collectProtocolFees(address(protocolFeeController), currency1, 0);
```
Comment (zhongeric) [inline test/foundry-tests/Fees.t.sol:626]: Shouldn't we be collecting `expectedProtocolFees ` here and not 0? I'm surprised the assertion on the next line passes

File: test/foundry-tests/Fees.t.sol [hunk 3]
```diff
@@ -581,6 +594,47 @@ contract FeesTest is Test, Deployers, TokenFixture, GasSnapshot {
...
+        assertEq(currency1.balanceOf(address(protocolFeeController)), expectedProtocolFees);
+    
+        uint256 expectedHookFees = 5; // 20% of 27 (30-3) is 5.4, round down is 5
+        vm.prank(address(hook));
+        // Addr(0) recipient will be the hook.
+        manager.collectHookFees(address(hook), currency1, 0);
```
Comment (zhongeric) [inline test/foundry-tests/Fees.t.sol:632]: Same as above

**Review by zhongeric — COMMENTED** [review]
Thanks for the contribution! Just some questions about a test but otherwise looks good to me!

**Comment by zhongeric** [comment]
I'll go ahead and merge in the latest main but all of your old commits also have to be signed

File: test/foundry-tests/Fees.t.sol (hunk 2 above)
Comment (sm-stack) [reply test/foundry-tests/Fees.t.sol:626]: At [L#89](https://github.com/Uniswap/v4-core/blob/9b912322d1456b9b2a7abd2585fd430ce19a12c5/contracts/Fees.sol#L89) `contracts/Fees.sol`, if the `amount` is zero, the function collects all protocol fees accrued. So the next `assert` should not generate error.

File: test/foundry-tests/utils/MockInvalidERC20.sol (hunk 1 above)
Comment (sm-stack) [reply test/foundry-tests/utils/MockInvalidERC20.sol]: Right, fixed it!

**Comment by sm-stack** [comment]
> Could you merge in the latest main to your branch and make sure you sign all of your old commits? Thanks!

Merged and fixed what you've mentioned!

**Comment by snreynolds** [comment]
Hi thanks for the contribution. I think [this PR](https://github.com/Uniswap/v4-core/pull/372) actually closed this change so I am going to close. However, there is still an outstanding tests to migrate the gas snapshots in PoolManager.gas.spec if you wanted to take that on!

**Pull request closed without merging**